docs/source/enterprise-guide/graph-security.mdx

```diff
@@ -0,0 +1,262 @@
+
+#### Only allow the gateway to query subgraphs directly
+
+As a best practice for federated data graphs, Apollo recommends that subgraphs services may only be queried by the gateway Apollo Server directly. The [Apollo Federation specification](https://www.apollographql.com/docs/federation/federation-spec/) outlines that subgraph schemas will have `_entities` and `_service` root fields on the `Query` type to assist with composition and query planning. An `sdl` field may be queried in the `_service` selection set and the output of that field will include the SDL representation of the subgraph's schema. Because this query will expose as much data about a subgraph's schema as a standard introspection query would, you should ensure this information isn't accessible to the outside world.
+
+Another reason to restrict access to subgraphs is related to the collection of field-level traces. This tracing data is included in the `extensions` key of the response from a subgraph to the gateway where the data is aggregated into a trace shape based on the query plan and then sent to Apollo Studio. That means that any client that can query your subgraphs directly would be able to view this data in the operation response and make inferences about a subgraph based on it.
+
+Lastly, beyond these security concerns, preventing the outside world from accessing subgraphs directly also helps you ensure that clients (even well-meaning ones) route operations to the consolidated data graph only and don't invent unintended use cases for the types and fields in a subgraph schema that are solely meant for executing the gateway's query plan.
+
+### Malicious queries
+
+Once there are measures in place to limit API discoverability in public-facing environments, the next step in protecting a GraphQL API is to guard it against both intentionally and unintentionally malicious queries. Again, many GraphQL-related vulnerabilities have to do with how an unprotected API may be exploited in DoS attacks, but there are other considerations as well. 
+
+In the sections that follow, we will explore measures that can help mitigate the impact of malicious queries for any GraphQL API, such as limiting query depth and amount, paginating list fields where appropriate, validating and sanitizing data, setting timeouts, and guarding against batched query abuse. And for GraphQL APIs with third-party clients, we will also explore using query cost analysis to support rate-limiting.
+
+#### Limit query depth
+
+GraphQL allows clients to traverse through a data graph and express complex relationships between the nodes in an operation's selection set. But as far as backing data sources are concerned, this can quickly turn into too much of a good thing when there are no guardrails to restrict how deeply queries can be nested. For example:
+
+```graphql
+query {
```

Comment:
Bit pedantic of me, but trying to align with our [operation best practices](https://www.apollographql.com/docs/react/data/operation-best-practices/) to always provide operation names. 😉 

```suggestion
query DeepBlogQuery {
```

Feel free to reject this!

Reply:
Added, thanks.